### ISSUE TITLE
feature - support Editors & Filters with Async Collections

### ISSUE THREAD

**Issue by ghiscoding** [ghiscoding/aurelia-slickgrid#83]
**I'm submitting a feature request**

**Current behavior:**
Since this is currently not available, user need to resolve collections **before** creating the grid

**Expected/desired behavior:**

* **What is the expected behavior?**
Simply pass an `http` Promise to an async collection property and let it do it's job

* **What is the motivation / use case for changing the behavior?**
Make it simpler to load asynchronously a collection and provide it to filter/editor without delaying the grid creation. 

#### TODO
- [x] Proof of Concept
- [x] Add a JSON file to simulate real `HttpClient` call for `collectionAsync`
- [x] Cleanup the code in MultipleSelectFilter
- [x] Extract duplicate functions/code in `MultipleSelectFilter`, `SingleSelectFilter` & `SelectFilter` and create a shared utility file or class
- [x] implement the same changes to the Editors `MultipleSelectEditor`, `SingleSelectEditor`
- [ ] update [Editor Wiki](https://github.com/ghiscoding/aurelia-slickgrid/wiki/Editors) & [Filter Wiki](https://github.com/ghiscoding/aurelia-slickgrid/wiki/Select-Filter)

**Comment by ghiscoding** [comment]
I now know how to approach this feature and will address it soon. We will have a new property inside editor/filter that will be named `asyncCollection` and is meant to load and create/recreate the editor/filter once the collection becomes available. 

For example
```javascript
{ 
   id: 'duration', name: 'Duration (days)', field: 'duration', 
    filter: {
      asyncCollection: new Promise<any>((resolve) => {
        // prepare a multiple-select array to filter with
        const multiSelectFilterArray = [];
        for (let i = 0; i < 100; i++) {
          multiSelectFilterArray.push({ value: i, label: i });
        }

        // simulate async server load
        setTimeout(() => {
          resolve(multiSelectFilterArray);
        }, 500);
      }),
      model: Filters.multipleSelect,
      searchTerms: [1, 33, 50], // default selection
    }
},
```

**Comment by jmzagorski** [comment]
I was thinking about solutions to this some time ago for my editors, but found a workaround and never revisited. My workaround was to create a property observer on `c.params` (still on the old version) and invalidate the grid when it changed. Just an FYI on what I was doing:

```javascript
      if (this.columns && this.columns.length) {
        for (const c of this.columns) {
          // watch collection changes
          if (c.params && c.params.collection) {
            this.bindingSubscriptions.push(
              this._binder.propertyObserver(c.params, 'collection')
                .subscribe((newVal, oldVal) => this.grid.invalidate())
            );
          }
        }
      }
```

So this is definitely a great feature to add.

Also, i still have no pushed my app to production because i seem to be keeping push and pulled in every other direction not related to my project. As soon as i do i plan on finishing up issues here.

**Comment by jmzagorski** [comment]
The collection is a bindable property on my custom element, though the one downside is that it does not listen to "pushes" and "removals" from the array, so I am forced to replace the entire array (which is fine for me because i use redux)


```javascript
import { bindable } from 'aurelia-framework';
import { Formatters, Editors, FieldType } from 'aurelia-slickgrid';

export class RelationGridCustomElement {
  // bindable props
  @bindable elements = [];

  bind() {
    this.columns = [{
      editor: Editors.singleSelect,
      field: 'elementId',
      formatter: Formatters.collection,
      id: 'elementId',
      name: 'Name',
      params: {
        collection: this.elements,
        collectionSortBy: {
          property: 'displayName'
        },
        customStructure: {
          label: 'displayName',
          value: 'id'
        }
      },
      sortable: true,
      type: FieldType.number
    }
  }

  // this will fire the observer everytime the collection changes
  elementsChanged(newVal, oldVal) {
    this.columns.find(c => c.id === 'elementId').params.collection = newVal;
  }
}
```
I just wanted to show you how I was working with async collections in my app. My way might not be better, but at least you can see another way in case you run into a road block.

**Comment by jmzagorski** [comment]
and the view model that feeds this custom element also stores the `elements` as a property that gets updated anytime the collection needs to be updated

**Comment by ghiscoding** [comment]
what is the `elements` array? is that entire list of elements that you can use to edit and that if you add a row to the grid, this array will grow by 1, is that it?

My goal, at least with the Filter that I know it works, is to use a Promise and recreate the DOM element once it resolves (typically a WebAPI call to get the collection). This is working fine for a Filter since we create it only once. But then after what you wrote, I realize that if I do this trick every single time I click on an Editor, it will call the Promise (WebAPI) as many time as I open an Editor which is not viable on the long run. The other thing I could maybe do is keep the Promise inside the `asyncCollection` and once it resolves then store it in the `collection` property for the entire life of the grid, this way I don't need to run the Promise if I know it was already resolved once. However, that won't work for people who want to refresh the collection, anytime a new row is added... or actually, any user could potentially refresh the `collection` again, whenever they feel like it, that might work just enough, which is essentially what you're doing I think!?

Oh and as per your first post, I totally understand being pulled away from the project. I myself, got pulled away by this grid for the past 6 months haha... but that's ok, I wanted to push this grid to another level and I achieved almost everything I wanted so I can slowly get back to my personal project, which was basically to build something with Aurelia/NodeJS/RethinkDB/Socket.IO and I needed a grid so it's all good hehe... and I made some good friend here ;)  I like that you give me another point of view, we all think differently :D

**Comment by jmzagorski** [comment]
thanks! it has definitely been a pleasure working on this library with you, I have learned a bunch. I agree about a different point of view. Most of the time I am developing by myself and it can be frustrating.

Yes that is what i am doing. The `elements` array is the entire list of elements they can select in an editor column. An element can be dynamically added to the editor and it will be show in the drop down as shown below when I set a timeout to add "AAA" (although if the editor is open as in the example below i am force to reclick the editor).

![2018-07-13_22-42-11](https://user-images.githubusercontent.com/8404957/42720040-2c3c99a8-86ee-11e8-81b5-782f16c251d7.gif)

**Comment by ghiscoding** [comment]
Yeah that is probably because we are recreating the DOM element every time. And also because, as you said, you are looking for any changes without knowing if it's adding, deleting or completely replacing the collection, so in that case we have no choice but to recreate the entire element. 
> (although if the editor is open as in the example below i am force to reclick the editor).

Altough, if we look at adding a new option to the multiple select, by looking at the site doc, it seems that we can append an option without recreating the DOM element as shown [here](http://wenzhixin.net.cn/p/multiple-select/docs/#the-refresh) 

```javascript
const value = 'new text';
$opt = $("<option />", {
                    value: value,
                    text: value
                });
$select.append($opt).multipleSelect("refresh");
```

However, dealing with adding/deleting/replacingCollection brings a bigger complexity and I would probably just opt for recreating the DOM element every time for now. We can always refine it later. 

For the refinement of knowing if it's adding/deleting/replacing, we can use `bindingEngine.collectionObserver` like I did in the lib itself [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/aurelia-slickgrid.ts#L266)

**Comment by ghiscoding** [comment]
So I tried it on a Filter and it all works with the `bindingEngine`, I can detect the type of change with `bindingEngine` and I decided to use the 2 line of code that I shown in previous post. The only problem that I see with adding the option to the `multipleSelect` is that I can `append` or `prepend` but I won't know which one to use, neither would I know if it's really suppose to be at the beginning/end of this select. So I might end up recreating the DOM element anyway since the collection will always be in the order that the user defined, that is untouched/sorted/filtered. 

Since it works nicely on a Filter, I will probably replicate the code with Editor

Anyhow, here's the code with a `prepend` (in the animated GIF below, I add new line 501, also note that because of my high res screen, the mouse cursor is offset)

```javascript
  private async renderOptionsAsync(asyncCollection: Promise<any>) {
    if (asyncCollection) {
      // wait for the "asyncCollection", once resolved we will save it into the "collection" for later reference
      const awaitedCollection: any[] = await asyncCollection;
      this.columnDef.filter.collection = awaitedCollection;

      // recreate Multiple Select after getting async collection
      this.renderDomElement(awaitedCollection);

      // subscribe to the "collection" property changes with BindingEngine
      this.subscriptions.push(
        this.bindingEngine
          .collectionObserver(this.columnDef.filter.collection)
          .subscribe((changes: { index: number, addedCount: number, removed: any[] }[]) => {
            const updatedCollection = this.columnDef && this.columnDef.filter && this.columnDef.filter.collection || [];
            if (Array.isArray(changes)) {
              changes.forEach((change) => {
                const option = updatedCollection[change.index];
                if (change.addedCount) {
                  // add the new item to the Multiple Select
                  const $opt = $('<option />', { value: option.value, text: option.label });
                  this.$filterElm.prepend($opt).multipleSelect('refresh');
                } else if (Array.isArray(change.removed) && change.removed.length > 0) {
                  // instead of trying to find the deleted item in the Multiple Select
                  // it's easier to just recreate the Select DOM Element
                  this.renderDomElement(updatedCollection);
                }
              });
            }
          })
      );
    }
  }
```
![2018-07-14_16-10-23](https://user-images.githubusercontent.com/643976/42728150-d6293974-8781-11e8-8c82-4858e1f8eb43.gif)

**Comment by ghiscoding** [comment]
@jmzagorski 
I'd like some feedback on previous post and I'm also thinking about creating a utility file to extract common functions of (SelectFilter, SingleSelectFilter, MultipleSelectFilter) but I'm wondering if I should create a Class or just functions? The reason I'm unsure is that a lot of these functions requires stuff from `this.args`, `this.columnDef`, `this.$filterElm`, ... so I'm unsure on how to approach that refactoring

**Comment by jmzagorski** [comment]
@ghiscoding I was able to review the change this morning and the logic looks good for the select filter. I do have 1 observation

You may not need to use promises in the filter or add the `asyncCollection`. Since you are already using `bindingEngine` you can use `propertyObserver` to track changes to the `collection` to update the dom element. For example, you can use this in the filter logic.

```typescript
    this.bindingEngine.propertyObserver(this.columnDef.filter, 'collection')
      .subscribe((newVal, oldVal) => this.renderDomElement(newVal));
```

With that code you will be watching for changes and your `collectionObserver` will watch for pushes and removals to the collection. If you try that out in your branch you will see you do not need promises. I added a simply `setInterval` to example 4 that changes the entire collection every two seconds and the select list updated properly.

**Comment by ghiscoding** [comment]
Oh that is interesting I didn't know we could add a 2nd argument for the property name. I'm a bit surprised on the return though, I always saw it returned a `changes` and not a `(newVal, oldVal)` but I can give that a try. 

Note I'm renaming `asyncCollection` to `collectionAsync`, the name is better and it makes them closer together and are best friends lol

As for the Promise, I'm not sure if this is what you mean but here... I might not have explained well what it does. Currently in the code example it's a Promise but in real code it will be an `HttpClient` call which I believe returns a Promise. Real code would look like this
```javascript
{
    id: 'duration', name: 'Duration (days)', field: 'duration', 
    filter: {
      collectionAsync: this.http.fetch('api/duration-list'),
      model: Filters.multipleSelect
   }
},
``` 
then if the user wants to know the current list used in the Filter, he can poke the `collection`, same goes for adding/removing options, he will use the `collection`. I could create another JSON file to make a real Http Fetch call, that might be more obvious for the example.

This is probably a better explanation that I can put in the Wiki when we get it all working. 
BTW I don't think I'll do much work this weekend, going to visit my mum and enjoy the sun hehe

**Comment by jmzagorski** [comment]
Well enjoy not working! Hopefully I will not be doing much either.  I do better understand `collectionAsync` looking at your example, though I still think you should add the `properyObserver` too
for users not using Promises or for users replacing the collection more than once. `collectionAsync` will only support promises and only update the entire collection once after it is resolved. If the users replaces the collection, the filter will never know about it. `propertyObserver` will work for any asynchronous programming (promises, generators, timeouts, intervals etc) and continue to replace the entire collection as long as the user changes it. Here is a timeout in example4 that clears the collection after 2 seconds that will not work without `propertyObserver`:

```typescript
 attached() {
    // populate the dataset once the grid is ready
    this.dataset = this.mockData(NB_ITEMS);

    setTimeout(() => {
      const durationColumnDef = this.columnDefinitions.find((column: Column) => column.id === 'duration');
      console.log("cleared");
      durationColumnDef.collection = [];
    }, 2000)
  }
```

`propertyObserver` and `collectionObserver` do different things as I am sure you know. `collectionObserver` will watch changes to the collection (`push`, `unshift`, `splice` etc.). `propertyObserver` will only watch when the entire collection is set to a new collection, which is why you get `newVal` and `oldVal` since the entire property changed.

**Comment by ghiscoding** [comment]
oh I didn't noticed you had put the `propertyObserver` instead of the `collectionObserver`, now I understand how you got the `newVal` and `oldVal` :P

I know that the `collectionAsync` is run once, it's mostly there for a convenience of adding an Http Fetch about it, which is what most people will do. For the rest of the other functionalities then user will continue to use the `collection` directly, we can surely have both type of observers. I actually didn't think about replacing the collection entirely, so I guess that I assumed wrongly that the `collectionObserver` would also know if it's being replaced. I kinda learn something new, never thought of using both observers to do different things but it make sense when you think about it... and this is where Angular suck, they used to have a `$watchCollection` in AngularJS but they don't have that functionality in 2+, so I need to use an observable (RxJS) and that is quite annoying, not only do I need to push on the `collection`, but I also need to trigger the Observable, so Aurelia win big time here

**Comment by jmzagorski** [comment]
very cool. Nice work!

**Comment by ghiscoding** [comment]
@jmzagorski 
Currently both observers are attached to the `collection` only on Filters (also Editors in future) that have a `collectionAsync` property initialized (with Promise or HttpClient) but I was wondering if I should add these observers on all Editors/Filters `collection` (no matter if you used or not the `collectionAsync`). 

Does that make any sense or will that possibly add too much overhead with performance hit because they will be as much observers as there are Editors/Filters. What do you think? 

Hmmm after some thoughts, that might be too much overhead. In most cases they won't be used, but I'm not sure the user will always have a `collectionAsync` at first. He might still want observers without providing an initial `collectionAsync`... unless we provide an extra flag for these cases. For example, if you add a new row that add a new option to an Editors, you would probably want the `collection` to be updated even though you didn't use `collectionAsync` at first since you had them all on hand.

**Comment by jmzagorski** [comment]
I think you should add them to the collection no matter what and here is my logic:

1. I do not think it will be that much overhead for editors because editors are only in use one at a time. Also, you might get away with not observing in every editor if you did not want to and the only downside is that the collection wont update when the editor is active (you have to ask yourself how often will the collection update when the user is using that select list?) The good thing about editors is that it only inspects the collection when the editor is activated, so the library could observe the collection on each columnDef in the AureliaSlickgridCustomElement and refresh the slickgrid columns through `setColumns`.
2. I am unsure about Filter overheard because I am unsure how they behave. However, filters probably rarely need to re-render themselves as you are scrolling through slickgrid. 
3. I know from the perspective of someone using redux, or the new aurelia-store, they would not use `collectionAsync` and only `collection` because every property in their view model will be updated via an observed pattern, and they would be injecting the `store` as not a `fetch` client such as the `HttpFetchClient`

**Comment by ghiscoding** [comment]
Thanks for the feedback

I'm not sure about this paragraph in number 1, especially that last few words to use `setColumns`.
> The good thing about editors is that it only inspects the collection when the editor is activated, so the library could observe the collection on each columnDef in the AureliaSlickgridCustomElement and refresh the slickgrid columns through `setColumns`.

2. Right for Editors, the user simply need to update the `collection` and the Editor will contain all of it every time. So I might not need to do anything for them then

3. Yes I can imagine and the `collectionAsync` is really meant to be used only once and mainly with an Http call which is on the page init. After that, you simply change the `collection` directly. However in Angular, that's another story since they don't have collection observer, so I use `collectionAsync` to keep the Observable and the user has to use that Observable to call an `emit` (or `next`). Thanks to Aurelia with their observers, I don't have to do the extra work like I do in Angular. 

I was thinking about it last night and at the time, I decided that it was better under a flag ( I can still remove it) and the flag is the following. However as you mentioned, we might not need any observers in the Editors, just update the `collection` directly, so I might remove the new flag since it's only for Filter.

```javascript
  /**
   * Defaults to false, when enable it will add collection observers and re-render the Filter DOM element
   * with the new collection when changes are detected. Also note that using "collectionAsync" automatically watch for changes,
   * in consequence, there's no need to enable this flag in that particular case.
   */
  enableCollectionWatch?: boolean;
```

**Comment by jmzagorski** [comment]
As for the paragraph you are not sure about, i think you sort of understood what I was saying, but here is more clarification:

I do not think the editor collection will update on the grid when a users updates the `collection` in their view model because slickgrid uses `extend` to create new column objects. I am pretty sure you have to call `setColumns` on the slickgrid object to refresh any propery on the columnDef .

**Comment by ghiscoding** [comment]
oh no I didn't understand it that way. I forgot that Slickgrid uses `extend` for the columns so yeah I see what you mean now. Filters are created by my lib not by Slickgrid which is why it works. Now I understand, thanks for the clarification.

**Comment by ghiscoding** [comment]
Some quick research about what you said, I just found this [jsfiddle](http://jsfiddle.net/origineil/ckkjy90b/) and the person uses the `metadata` to change the editor

```javascript
data.getItemMetadata=function(row) {    
        var rowData = data[row]
        //console.log(rowData)
        var metaData = {columns:{other: {}}}
        metaData.columns.other.editor = Slick.Editors[rowData.type]         
        return metaData
}
```
in this example, `other` is a column name.

So will have to check if it works with `setColumns` and/or the `metadata`

**Comment by ghiscoding** [comment]
OMG I finally got it working (Editor async) after spending like 8 hours on this thing. I tried so many things that I was about to throw the computer through the window lol (though it's the work computer, so I probably shouldn't) 

So here's a quick recap
1- the `metadata` doesn't work so good, I have a hard time to deal with it and doesn't work in our scenario because it only updates on a click even but the click is before not after (so anyway can't use that). 
2- use the `setColumns` to reassign since SlickGrid internally use `extend`, well it's actually not entirely true. I mean for some reason, I do need to use the `getColumns` because it's async (and also bcoz user might have moved the column with `presets`), however using the reference pointer and reassigning the `collection` to the `internalColumnEditor` is enough.
3- Lastly I was trying to do the changes inside the Editor but actually we loop through the editors inside Aurelia-Slickgrid, so that is where I need to do my stuff.

@jmzagorski 
If you see anything strange, let me know

Here's a quick summary of the code (note that Angular uses RxJS Observables, not Promises for HttpClient, I hopefully will be able to simply use collection observers in Aurelia). 

```javascript
bind( ) {
    this._columnDefinitions = this._columnDefinitions.map((column: Column | any) => {
      if (column.editor && column.editor.collectionAsync) {
        this.loadEditorAsyncCollection(column);
        this.createNewCollectionAsyncObservable(column);
      }
      return { ...column, editor: column.editor && column.editor.model, internalColumnEditor: { ...column.editor  }};
    });
}

protected createNewCollectionAsyncObservable(column: Column) {
    if (column && column.editor) {
      // create new Subject, then reassign as the new collectionAsync and finally watch for changes on it
      const newCollectionAsync = new Subject<any>();
      column.editor.collectionAsync = newCollectionAsync;
      this.subscriptions.push(
        newCollectionAsync.subscribe(updatedCollection => this.updateInternalEditorCollection(column, updatedCollection))
      );
    }
  }

private loadEditorAsyncCollection(column: Column) {
    const collectionAsync = column && column.editor && column.editor.collectionAsync;
    if (collectionAsync instanceof Observable) {
      this.subscriptions.push(
        collectionAsync.subscribe((resolvedCollection) => this.updateInternalEditorCollection(column, resolvedCollection))
      );
    }
  }

private updateInternalEditorCollection(column: Column, newCollection: any[]) {
    column.editor.collection = newCollection;
    column.internalColumnEditor = column.editor;

    // find the new column reference pointer
    const columns = this.grid.getColumns();
    if (Array.isArray(columns)) {
      const columnRef: Column = columns.find((col: Column) => col.id === column.id);
      columnRef.internalColumnEditor = column.editor;
    }
  }
```

The `createNewCollectionAsyncObservable` is not necessary in Aurelia since we can use Observers (which Angular doesn't seem to have... Aurelia wins).

and here's an animated GIF of the Editor Example with both Filters & Editors being loaded asynchronously (using `collectionAsync`) then using Observable (or Observers in Aurelia) when adding new item(s).
![2018-08-07_18-46-23](https://user-images.githubusercontent.com/643976/43806499-3af0d288-9a72-11e8-9906-c1d1897c37e0.gif)

**Comment by ghiscoding** [comment]
@jmzagorski 
Just saw your name in the Aurelia blog post, nice job ;)

> Besides the update to RxJS 6, we've improved the connectTo decorator, which with the help of @jmzargorski, now has support for multiple selectors. This opens up a whole new workflow in working with multiple modules and making parts of your app even more self-contained in regards to the Store
